python/mlx/nn/layers/gnn.py

```diff
@@ -0,0 +1,253 @@
+from typing import Optional
+
+import mlx.core as mx
+from mlx.nn.layers.base import Module
+
+
+class GraphNetworkBlock(Module):
+    """Implements a generic Graph Network block as defined in [1].
+
+    A Graph Network block takes as input a graph with N nodes and E edges and
+    returns a graph with the same topology.
+
+    The input graph can have:
+    - `node_features`: features associated with each node in the graph, provided as an
+        array of size [N, F_N]
+    - `edge_features`: features associated with each edge in the graph, provided as an
+        array of size [E, F_E]
+    - `global_features`: features associated to the graph itself, of size [F_U]
+    The topology of the graph is specified as an `edge_index`, an array of size [2, E],
+    containing the source and destination nodes of each edge as column vectors.
+
+    A Graph Network block is initialized by providing node, edge and global models (all
+    optional), that are used to update node, edge and global features (if present).
+    Depending on which models are provided and how they are implemented, the Graph
+    Network block acts as a flexible ``meta-layer`` that can be used to implement other
+    architectures, like message-passing networks, deep sets, relation networks and more
+    (see [1]).
+
+
+    Args:
+        node_model (mlx.nn.layers.base.Module, optional): a callable Module which updates
+            a graph's node features
+        edge_model (mlx.nn.layers.base.Module, optional): a callable Module which updates
+            a graph's edge features
+        global_model (mlx.nn.layers.base.Module, optional): a callable Module which updates
+            a graph's global features
+
+    References:
+        [1] Battaglia et al. Relational Inductive Biases, Deep Learning, and Graph Networks. https://arxiv.org/pdf/1806.01261.pdf
+
+
+    .. code-block:: python
+        import mlx.core as mx
+
+        from mlx.nn.layers.linear import Linear
+        from mlx.nn.layers.gnn import GraphNetworkBlock
+
+
+        class NodeModel(Module):
+            def __init__(
+                self,
+                node_features_dim: int,
+                edge_features_dim: int,
+                global_features_dim: int,
+                output_dim: int,
+            ):
+                super().__init__()
+                self.model = Linear(
+                    input_dims=node_features_dim + edge_features_dim + global_features_dim,
+                    output_dims=output_dim,
+                )
+
+            def __call__(
+                self,
+                edge_index: mx.array,
+                node_features: mx.array,
+                edge_features: mx.array,
+                global_features: mx.array,
+            ):
+                destination_nodes = edge_index[1]
+                aggregated_edges = mx.zeros([node_features.shape[0], edge_features.shape[1]])
+                for i in range(node_features.shape[0]):
```

Comment:
This could be very slow for 'large' graphs.

Neighborhood aggregation is a very common operation in GNNs which can be implemented more efficiently using ``index_select`` and ``scatter_sum``.  It might be useful to implement an efficient ``index_select`` and ``scatter_sum`` first and then utilize them to implement neighborhood aggregation.

Reply:
Yes, this is definitely very slow and one of the reasons why it's just implemented in the docstring example and not as an actual class.

`scatter` operations seems to be all implemented in C++ (https://github.com/ml-explore/mlx/blob/608bd43604b05000a4401a9adffa383ca68618f6/mlx/ops.cpp#L2183) but as you say there seem to be no python bindings.

Reply:
I think we should create a PR to expose ``scatter``.

Reply:
Happy to collab on that.

Reply:
Here is a recent PR where I added Python bindings to C++ implementation (https://github.com/ml-explore/mlx/pull/187). I am currently working on pooling layers. I can work on ``scatter`` bindings tonight. If you need it earlier, please  create the initial version of PR :)

Reply:
Regardless of whether GNN gets merged into MLX, having ``scatter`` in Python seems useful.

Reply:
@gboduljak created initial PR (https://github.com/ml-explore/mlx/pull/391) to collaborate on. Any help welcome!